### ISSUE TITLE
spire-agent fails to start because trust bundle is not available

### ISSUE THREAD

**Issue by szvincze** [spiffe/spire#3032]
* **Version**: 1.1.0, 1.2.0, 1.2.1, 1.2.2, 1.2.3
* **Platform**: Linux 5.15.0-27-generic#28-Ubuntu SMP x86_64
* **Subsystem**: agent

Scenario:
The spire server and agents are launched at once.

After spire-server is started properly and its TCP server is established, spire-agent's initContainer detects the server port is open thus it launches the normal agent container, which fails with the following error message:
```could not parse trust bundle: open /run/spire/bundle/bundle.crt: no such file or directory```

It seems there is a race condition between the spire-server that creates the trust bundle and spire-agents which would like to parse it. Thus after spire-agent throws the above error kubelet restarts the container and usually after one restart the trust bundle is in place and the spire-agent gets up and running.

```
$ kubectl get pods -nspire --watch
NAMESPACE NAME                 READY   STATUS              RESTARTS   AGE
spire     spire-server-0       0/2     Pending             0           0s
spire     spire-server-0       0/2     Pending             0           0s
spire     spire-server-0       0/2     ContainerCreating   0           0s
spire     spire-agent-q2h7l    0/1     Pending             0           0s
spire     spire-agent-q2h7l    0/1     Pending             0           0s
spire     spire-agent-sdlh6    0/1     Pending             0           0s
spire     spire-agent-sdlh6    0/1     Pending             0           0s
spire     spire-agent-q2h7l    0/1     Init:0/1            0           0s
spire     spire-agent-sdlh6    0/1     Init:0/1            0           0s
spire     spire-agent-sdlh6    0/1     Init:0/1            0           2s
spire     spire-agent-q2h7l    0/1     Init:0/1            0           2s
spire     spire-server-0       1/2     Running             0           5s
spire     spire-server-0       2/2     Running             0           5s
spire     spire-agent-q2h7l    0/1     PodInitializing     0           6s
spire     spire-agent-sdlh6    0/1     PodInitializing     0           7s
spire     spire-agent-q2h7l    0/1     Running             0           7s
spire     spire-agent-sdlh6    0/1     Error               0           8s
spire     spire-agent-sdlh6    0/1     Running             1 (2s ago)  9s
spire     spire-agent-sdlh6    1/1     Running             1 (8s ago) 15s
spire     spire-agent-q2h7l    1/1     Running             0          15s
```

As confirmed in the https://github.com/spiffe/spire/issues/2978, opened by @richardstone, there is no other implication than the container restart, but the checks of the restart counters in the different verification loops make the tests fail.

For some reason the situation remains the same even if I added few seconds sleep in the wait-for-it init container:
```args: ["-t", "30", "spire-server:8081", "--", "sleep", "5"]```

When I added check and waiting for the trust bundle in the init container before starting the spire-agent container it resulted more than 70 seconds delay to get the running status, however the restart did not come. So, it seems to wait for the bundle in the init container does not solve the problem, at least not properly.

**Comment by amartinezfayo** [comment]
> When I added check and waiting for the trust bundle in the init container before starting the spire-agent container it resulted more than 70 seconds delay to get the running status

Just want to confirm that you are referring to the fact that the init container waits for the bundle to be present in the volume before starting the agent.

In #2978 it seems that the problem was because there were multiple SPIRE servers with their own datastore. Do you have one or more servers? In case of multiple servers, how the datastore is configured?

**Comment by amartinezfayo** [comment]
I'm thinking that there may be a k8s configuration related with [how ConfigMaps are updated](https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically), that is producing this behavior. In particular, I'm wondering if this situation is influenced by the [`configMapAndSecretChangeDetectionStrategy`](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) setting.
> A ConfigMap can be either propagated by watch (default), ttl-based, or by redirecting all requests directly to the API server. As a result, the total delay from the moment when the ConfigMap is updated to the moment when new keys are projected to the Pod can be as long as the kubelet sync period + cache propagation delay, where the cache propagation delay depends on the chosen cache type (it equals to watch propagation delay, ttl of cache, or zero correspondingly).

In any case, it's hard for me to think that this is caused by SPIRE itself. I'm more inclined to think that there is some configuration in the environment that is causing this.

@szvincze could you please check what's the configured ConfigMap change detection strategy in your deployment?

**Comment by szvincze** [comment]
> @szvincze could you please check what's the configured ConfigMap change detection strategy in your deployment?

@amartinezfayo I have checked my environments and found no value configured for this attribute. Thus I assume the default value "Watch" is used in this case.

Could you please help me to understand how it is supposed to work?

For me it looks like there is a timing issue because spire-server and agent are started in the same time, as helm does this way. Even if spire-agent is waiting in init for the server's port, after that it tries to parse the bundle file that is not there yet. However it can happen that I misunderstand something.

**Comment by szvincze** [comment]
> SPIRE Agent does not have any logic to determine when the trust bundle has been populated, it will always assume that it's there. I think that syncing when this resource is available to be consumed should not be a responsibility of SPIRE, it feels more tied to the tuning that you can do with the container orchestrator.

There is an init container (wait-for-it) which is waiting for the SPIRE Server's TCP port to be open. My understanding is that the trust bundle is also a prerequisite for SPIRE Agent to be able to start up and get ready status.
I have tried to implement a check in the init container that is waiting for the server port but that was not really successful (as @richardstone also indicated in issue #2978, because the delay of the agent pod's start was more than 70 seconds.

That is why I wanted to know if there is any dependency between the agent and server from the bundle creation point of view.

Do you see any explanation of this behavior?

**Comment by szvincze** [comment]
I included a small [function](https://github.com/Nordix/spire/blob/3032/cmd/spire-agent/cli/run/run.go#L613-L626) in spire-agent that checks if the trust bundle file is accessible. It seems from the logs that in worst case the file gets in place within 200ms. (One waiting loop then the repeated one is successful.) However, as I mentioned above, waiting for the bundle in an init container does not work for some reason.

@amartinezfayo What is your opinion?

**Comment by amartinezfayo** [comment]
The explanation that I can think of is that the init container has a different view of the ConfigMap.

Isn't relying on agent restarts a possible solution?

**Comment by amartinezfayo** [comment]
We have discussed this issue in the last maintainer's sync, looking for different alternatives to solve this problem.
Since the timing between when the server publishes its bundle to a ConfigMap and when the agent container sees it is not under control of SPIRE, it doesn't seem that would be appropriate to implement a logic within SPIRE to accommodate that (like some retry logic in the agent).

I'm personally more inclined to find a solution implemented either in the init container or the agent container (e.g. the init container could wait for some time before returning a positive status).

**Comment by szvincze** [comment]
@amartinezfayo Thanks for your comment.

As I wrote previously, unfortunately it did not work when we included checking of the trust bundle file into the init container that waits for spire-server's TCP port to be open. After the port was available the second waiting loop started to check the bundle but that took more than a minute. However when I put this retry into spire-agent that took 0.2 second in worst case when it detected the bundle file.

Now I tried to create a second init container which did not do anything but wait 1 second, and it seems it is working fine this way. So, it looks like that for some reason when we would like to do both checks or waiting in one init container that does not work, but when the second check is done in another container, or at least some kind of interruption is needed between the two checks (TCP port and trust bundle).

So, it is a workaround that we can use but an event-driven solution would be preferable.
